src/Compilers/Core/Portable/Collections/UnionCollection.cs

```diff
@@ -68,6 +85,38 @@ private UnionCollection(ImmutableArray<ICollection<T>> collections)
             _collections = collections;
         }
 
+        private static ImmutableArray<ICollection<T>> Flatten(ImmutableArray<ICollection<T>> collections)
```

Comment:
@sharwell @CyrusNajmabadi do we have benchmarks on the impact that this has overall, not including simplify type names changes? /cc @agocke for running compiler benchmarks on a change.

Reply:
@sharwell do we really need the union-collection change?  i think it would be totally fine to just undo this bit for the purspose of this PR.  it seems unlikely we'd need to flatten deeply a tree of union collections.

the one operatoin we perform (`.Contains`) will be O(n) either way.

Reply:
The benchmarks I have did not include `Flatten`, so removing it would be fine.